mlir/lib/Driver/CompilerDriver.cpp

```diff
@@ -443,40 +463,54 @@ LogicalResult runLowering(const CompilerOptions &options, MLIRContext *ctx, Modu
                           CompilerOutput &output)
 
 {
+    using std::pair;
```

Comment:
I don't understand this function well. It looks too complicated. Would it be possible to do the following?

```c++
LogicalResult runLowering(const CompilerOptions &options, MLIRContext *ctx, ModuleOp moduleOp,
                          CompilerOutput &output)
{
    size_t pipelineIdx = 0;
    for (const auto &pipeline : options.pipelinesCfg) {
        auto pm = PassManager::on<ModuleOp>(ctx, PassManager::Nesting::Implicit);
        auto MLIRPipeline = joinPasses(pipeline.passes);
        auto diagnostic = options.diagnosticStream;
        if (failed(parsePassPipeline(MLIRPipeline, pm, diagnostic))) {
            return failure();
        }

        // TODO: Enable diagnostics

        // Run the lowering pipelines
        if (failed(pm.run(moduleOp))) {
            return failure();
        }
        
        // TODO: print the moduleOp
    }
}
```

Reply:
I think the reason it is complicated in its current stage is because a pipeline can contain nested pipelines. From MLIR's documentation.

```markdown
## Dynamic Pass Pipelines

In some situations it may be useful to run a pass pipeline within another pass,
to allow configuring or filtering based on some invariants of the current
operation being operated on. For example, the
[Inliner Pass](Passes.md/#-inline) may want to run
intraprocedural simplification passes while it is inlining to produce a better
cost model, and provide more optimal inlining. To enable this, passes may run an
arbitrary `OpPassManager` on the current operation being operated on or any
operation nested within the current operation via the `LogicalResult
Pass::runPipeline(OpPassManager &, Operation *)` method. This method returns
whether the dynamic pipeline succeeded or failed, similarly to the result of the
top-level `PassManager::run` method. A simple example is shown below:
```

Reply:
> Would it be possible to do the following?  (run mlir several times in a loop)

It might be possible, and would certainly simplify things. I didn't study the consequences. I think that in the design you are proposing we would need to take additional responsibility, e.g. we can not run passes in parallel (but I am not sure mlir can do this either).

Reply:
Passes are not run in parallel. But passes can have intra-pass multithreaded dispatch of patterns based on ops. Can you try the approach proposed above and have a commit to see how different the code would look like?

Reply:
@erick-xanadu I am glad to do this, but I would like to start this work as a different task. Could we limit the scope of this pr to just MT issues?

Reply:
FYI The [documentation](https://mlir.llvm.org/docs/PassManagement/) says that (1) several instances of a pass might run in parallel and (2) nested pipelines are supported. While it should be clarified, I think there are arguments towards keeping a single mlir pass-manager "program".

Reply:
You may be right, can you point to the specific section or quote that you read? Thanks!

Reply:
Ah, yes I see here:

>         Multiple instances of the pass may be created by the pass manager to process operations in parallel.